pkg/volume/downwardapi/downwardapi.go

```diff
@@ -0,0 +1,370 @@
+// TODO: Add volume specific ticker and refresh loop
+func (b *downwardAPIVolumeBuilder) SetUp() error {
+	return b.SetUpAt(b.GetPath())
+}
+
+func (b *downwardAPIVolumeBuilder) SetUpAt(dir string) error {
+	glog.V(3).Infof("Setting up a downwardAPI volume %v for pod %v/%v at %v", b.volName, b.pod.Namespace, b.pod.Name, dir)
+	// Wrap EmptyDir,let it do the setup.
+	wrapped, err := b.plugin.host.NewWrapperBuilder(wrappedVolumeSpec, b.pod, *b.opts, b.mounter)
+	if err != nil {
+		glog.Errorf("Couldn't setup downwardAPI volume %v for pod %v/%v: %s", b.volName, b.pod.Namespace, b.pod.Name, err.Error())
+		return err
+	}
+	if err := wrapped.SetUpAt(dir); err != nil {
+		glog.Errorf("Unable to setup downwardAPI volume %v for pod %v/%v: %s", b.volName, b.pod.Namespace, b.pod.Name, err.Error())
+		return err
+	}
+
+	data, err := b.collectData()
+	if err != nil {
+		glog.Errorf("Error preparing data for downwardAPI volume %v for pod %v/%v: %s", b.volName, b.pod.Namespace, b.pod.Name, err.Error())
+		return err
+	}
+
+	if !b.isDataChanged(data) {
+		// No data changed: nothing to write
+		return nil
+	}
+
+	if err := b.writeData(data); err != nil {
+		glog.Errorf("Unable to dump files for downwardAPI volume %v for pod %v/%v: %s", b.volName, b.pod.Namespace, b.pod.Name, err.Error())
+		return err
+	}
+
+	glog.V(3).Infof("Data dumped for downwardAPI volume %v for pod %v/%v", b.volName, b.pod.Namespace, b.pod.Name)
+	return nil
+}
+
+// IsReadOnly func to fullfill volume.Builder interface
+func (d *downwardAPIVolume) IsReadOnly() bool {
+	return true
+}
+
+// collectData collects requested downwardAPI in data map.
+// Map's key is the requested name of file to dump
+// Map's value is the (sorted) content of the field to be dumped in the file.
+func (d *downwardAPIVolume) collectData() (map[string]string, error) {
+	errlist := []error{}
+	data := make(map[string]string)
+	for fieldReference, fileName := range d.fieldReferenceFileNames {
+		if values, err := fieldpath.ExtractFieldPathAsString(d.pod, fieldReference); err != nil {
+			glog.Errorf("Unable to extract field %s: %s", fieldReference, err.Error())
+			errlist = append(errlist, err)
+		} else {
+			data[fileName] = sortLines(values)
+		}
+	}
+	return data, utilErrors.NewAggregate(errlist)
+}
+
+// isDataChanged iterate over all the entries to check wether at least one
+// file needs to be updated.
+func (d *downwardAPIVolume) isDataChanged(data map[string]string) bool {
+	for fileName, values := range data {
+		if isFileToGenerate(path.Join(d.GetPath(), fileName), values) {
+			return true
+		}
+	}
+	return false
+}
+
+// isFileToGenerate compares actual file with the new values. If
+// different (or the file does not exist) return true
+func isFileToGenerate(fileName, values string) bool {
+	if _, err := os.Lstat(fileName); os.IsNotExist(err) {
+		return true
+	}
+	return readFile(fileName) != values
+}
+
+const (
+	downwardAPIDir    = "..downwardapi"
```

Comment:
Comment on the crazy names?  Something like "It seems reasonable to allow dot-files in the config, so we reserved double-dot-files for the implementation".